File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
from itertools import combinations as comb
from itertools import permutations as perm
def solution(user, bann):
    ans = 0
    
    # 조합
    for c in comb(user,len(bann)):
        # 조합 속에서 하나의 순열이라도 일치하면 ans += 1
        for perms in perm(bann,len(bann)):
            f = True
            for u,b in zip(c,perms):
                # 길이가 안 맞으면 pass
                if len(u) != len(b):
                    f = False
                    break
                for i in range(len(u)):
                    if b[i] == '*': continue
                    if b[i] != u[i]:
                        f = False
                        break
            if f:
                # print(f"찾은 순열: {perms}")
                ans += 1
                break
    
    return ans

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.kt
class Solution {
    lateinit var combinations: List<MutableList<String>>
    val ans = mutableSetOf<String>()
    fun solution(user_id: Array<String>, banned_id: Array<String>): Int {
        val n = banned_id.size
        combinations = List(n) { mutableListOf<String>() }
        for((idx,ban) in banned_id.withIndex()) {
            for(user in user_id) {
                if (check(user,ban)) combinations[idx].add(user)
            }
        }
        dfs(0,mutableListOf(),mutableSetOf())
        return ans.size
    }
    fun check(user: String, ban: String): Boolean {
        if (user.length != ban.length) return false
        return user.zip(ban).all { (uc,bc) -> bc == '*' || uc == bc }
    }
    fun dfs(idx: Int, path: MutableList<String>, visit: MutableSet<String>) {
        if (idx == combinations.size) {
            val key = path.sorted().joinToString()
            ans.add(key)
            return
        }
        
        for (candidate in combinations[idx]) {
            if (candidate in visit) continue
            path.add(candidate)
            visit.add(candidate)
            dfs(idx+1,path,visit)
            path.removeLast()
            visit.remove(candidate)
        }
    }
}